README.md
# SIM_GENERATOR

A little command line utility to generate pattern images for structured illumination microscopy

## Installation

Head over to the [release](https://github.com/bradday4/SIM_GENERATOR/releases) page and download the latest executable `pattern_gen.exe`.

## Usage

### Minimum usage

From the directory that `pattern_gen.exe` is in

```Shell
pattern_gen <image_height> <image_width> <pattern_bar_width> --phase_shifts <n_shifts>
```

### Example with all arguments

```Shell
pattern_gen 256 256 16 --phase_shifts 3 --output_directory D:\SIM_GENERATOR\DemoPatterns --orientation horizontal --file_format jpg --bits 8 --full_depth True
```

Will generate the following images
|phi_01.jpg |phi_02.jpg |phi_03.jpg|
|--- |---|---|
|![](https://github.com/bradday4/SIM_GENERATOR/blob/main/DemoPatterns/phi_01.jpg) |![](https://github.com/bradday4/SIM_GENERATOR/blob/main/DemoPatterns/phi_02.jpg) |![](https://github.com/bradday4/SIM_GENERATOR/blob/main/DemoPatterns/phi_03.jpg)|

| Argument         | Type          | Optional? | Default                  | Options                     | Desc                                                                                                                         | Example                                    |
| ---------------- | ------------- | --------- | ------------------------ | --------------------------- | ---------------------------------------------------------------------------------------------------------------------------- | ------------------------------------------ |
| Positional Args  | (int int int) | False     | NONE                     | NONE                        | Image Height, Image Width, & Bar width                                                                                       | `1024 986 32`                              |
| phase shifts     | int           | False     | NONE                     | `0<n<inf`                   | How many phase shift images                                                                                                  | `--phase_shifts 5 `                        |
| output directory | str           | True      | `~exe location\patterns` |                             | where to save pattern images                                                                                                 | `--output_directory C:\Users\Thanos\Stuff` |
| orientation      | str           | True      | `vertical`               | `vertical` or `horizontal`  | which way should the bars be oriented                                                                                        | `--orientation vertical`                   |
| file format      | str           | True      | `bmp`                    | `bmp tiff tif jpg jpeg png` | File type                                                                                                                    | `--file_format tif`                        |
| bits             | int           | True      | `8`                      | `8 16 32`                   | Bit depth of image                                                                                                           | `--bits 16`                                |
| full depth       | bool          | True      | `True`                   | `True False`                | If true Use full range of bitdepth e.g 255 for `8` otherwise image will be 0 or 1. When `bits==32` option overides to`False` | `--full_depth True`                        |
| bar type         | str           | True      | `solid`                  | `solid triangle sine`       | Type of waveform                                                                                                             | `--bar_type sine`                          |

## Requests

If you have a feauture request open an issue and let me know.


CHANGELOG.md
# Changelog

<!--next-version-placeholder-->

## v1.3.8 (2021-07-19)
### Fix
* **auto release:** Artifact loads correctly to latest release ([`04698fe`](https://github.com/bradday4/SIM_GENERATOR/commit/04698fef082067978e90e1c172b58ef2febe7377))

## v1.3.7 (2021-07-19)
### Fix
* **auto release:** Upload workspace ([`67e5e70`](https://github.com/bradday4/SIM_GENERATOR/commit/67e5e707a8756730a405bfa143455e4f9df5c98a))

## v1.3.6 (2021-07-19)
### Fix
* **auto release:** Use github workspace path ([`c50b51c`](https://github.com/bradday4/SIM_GENERATOR/commit/c50b51c3aea5856780697a29cefa751f36dfe04f))

## v1.3.5 (2021-07-19)
### Fix
* **auto release:** Download artifact to second job ([`164fe43`](https://github.com/bradday4/SIM_GENERATOR/commit/164fe4371f1fec20bee1c8159446f466c37ea017))

## v1.3.4 (2021-07-19)
### Fix
* **auto release:** Artifact is folder not executable ([`437f017`](https://github.com/bradday4/SIM_GENERATOR/commit/437f0173876f066aae1074393b28060a57ba09bc))

## v1.3.3 (2021-07-19)
### Fix
* **auto release:** Change upload asset path ([`420fc25`](https://github.com/bradday4/SIM_GENERATOR/commit/420fc25a8c8d3bdc09a7220c87a729687b7ad696))

## v1.3.2 (2021-07-19)
### Fix
* **auto release:** Update path ([`2b793a9`](https://github.com/bradday4/SIM_GENERATOR/commit/2b793a983cd1e348fc9f0173d6d05192d0af77c0))
* **auto release:** Remove - from uses ([`1443570`](https://github.com/bradday4/SIM_GENERATOR/commit/144357097de3aa7a81bd502b298cec979a096166))
* **auto release:** Upload release asset ([`731d7db`](https://github.com/bradday4/SIM_GENERATOR/commit/731d7dbe83400c228894fff0e8f51f6853ce39e0))

## v1.3.1 (2021-07-19)
### Fix
* **auto release:** Build then deploy ([`01b767b`](https://github.com/bradday4/SIM_GENERATOR/commit/01b767bc4942bd180a4acd773770271ebf551431))

## v1.3.0 (2021-07-18)
### Feature
* **auto release:** Push artifact to current release ([`eff0b6c`](https://github.com/bradday4/SIM_GENERATOR/commit/eff0b6c2fceb572d59003386e504153d2c04c377))

### Fix
* **auto release:** Trying no newline ([`816a652`](https://github.com/bradday4/SIM_GENERATOR/commit/816a65203b718beb0d6271f73fc0ebd6c4b2d888))

## v1.2.2 (2021-07-14)
### Fix
* **release binary:** Change name to pattern_gen ([`7630ac0`](https://github.com/bradday4/SIM_GENERATOR/commit/7630ac044ec30ffb3b74243fbba32038f425770a))
* **release binary:** Build one file ([`8e95a54`](https://github.com/bradday4/SIM_GENERATOR/commit/8e95a54f381818a59172f449c33dec25c30cbd93))
* **release binary:** Remove root dir from spec file ([`229abce`](https://github.com/bradday4/SIM_GENERATOR/commit/229abcef5d78227714f7ba2286ceb9ef2aafe114))
* **release binary:** Custom spec file ([`d66e459`](https://github.com/bradday4/SIM_GENERATOR/commit/d66e459f10e68dbff32d86728df2ddb04b8d1fa0))
* **release binary:** Update spec file ([`f129a1a`](https://github.com/bradday4/SIM_GENERATOR/commit/f129a1a10ac6451e97031a26405512727b3dcbb4))
* **release binary:** Added spec file ([`bc7150e`](https://github.com/bradday4/SIM_GENERATOR/commit/bc7150e68bab80c1cfd71b7762f76f5204e4a746))

## v1.2.1 (2021-07-14)
### Fix
* **release binary:** Change packages to src ([`da9c745`](https://github.com/bradday4/SIM_GENERATOR/commit/da9c745f09de48927d98204aa294e807be1060db))

## v1.2.0 (2021-07-14)
### Feature
* **release binary:** Adding support for pyinstaller  to generate windows executable ([`4701005`](https://github.com/bradday4/SIM_GENERATOR/commit/4701005a5fe1802ea5d34a518dd2d00048acc70d))

## v1.1.1 (2021-07-13)
### Fix
* **python semantic release:** Hacky fix attempt for not recognizing install requires ([`0539ece`](https://github.com/bradday4/SIM_GENERATOR/commit/0539eceb785752e08cf252c5734f5dfe886f6826))

## v1.1.0 (2021-07-13)
### Feature
* **semantic-release-pypi:** Use built in pypi tool ([`7da8264`](https://github.com/bradday4/SIM_GENERATOR/commit/7da82649d6ab190330250c85350e20bd8e110e79))

### Fix
* **python semantic release:** Pull from main branch ([`acce3fa`](https://github.com/bradday4/SIM_GENERATOR/commit/acce3fa57fcea4d98f396bfbce759a2b3a86b407))
* **pyhthon semantic release:** Use main branch not master ([`f6040f3`](https://github.com/bradday4/SIM_GENERATOR/commit/f6040f39f4e344ebdfd338126f787af360c60764))
* **Python Semantic Release:** Using pure python implementation ([`bccd343`](https://github.com/bradday4/SIM_GENERATOR/commit/bccd343111735f3f0f1c11489c5c5c6f3a66fcb5))
* **semantic-release-pypi:** Add changelog support ([`c063e06`](https://github.com/bradday4/SIM_GENERATOR/commit/c063e0637f11ebbfd924f12120c501781f0e9ccf))
* **semantic-release-pypi:** Drop npm ci ([`9b69a97`](https://github.com/bradday4/SIM_GENERATOR/commit/9b69a973873a6e1d7cb1aa13867d5017045dc3f2))


square_wave.py
#%%
from skimage.io import imsave
from skimage import img_as_ubyte
from scipy import signal
import numpy as np
import matplotlib.pyplot as plt

# %%
def gen_pattern_image(signal, height):
    return np.tile(signal[np.newaxis, :], (height, 1))


def rescale(x, lims=(0, 1)):
    new_min, new_max = lims
    old_min, old_max = np.min(x), np.max(x)
    return (new_max - new_min) * ((x - old_min) / (old_max - old_min)) + new_min


# lh : number of vertical elements on the optical device
# lv : number of horizontal elements on the optical device

# SLM
# lh = 1536
# lv = 2048

# DMD
lh = 912
lv = 1140

wl = 4  # (width of line in pixels)
freq = lv / wl / 2
# phi_deg = [0.001, 72.0, -72.0, 60.0, -60.0] # non uniform
phi_deg = [0.001, 72.0, 144, 216, 288]  # uniform (can't handle 0 or 360)
# phi_deg = [0.000, 144.0, -144.0]  # non uniform
phi_rad = (np.radians(deg) for deg in phi_deg)  # phase in radian
t = np.linspace(0, 1, lv, endpoint=False)
ims = []
for phi in phi_rad:
    # ims.append(gen_pattern_image(signal.square(2 * np.pi * freq * t + phi), lh))
    x = np.sin(2 * np.pi * freq * t + phi)
    x = rescale(x)

    ims.append(gen_pattern_image(x, lh))
    # plt.plot(t * 2048, signal.square(2 * np.pi * freq * t + phi))
    # plt.ylim(-2, 2)


# %%
for i, im in enumerate(ims):
    imsave(
        r"C:\Users\Brad\Desktop\eq_sine\vsine_f{freq:.0f}pix_{i}.bmp".format(
            freq=wl, i=i
        ),
        img_as_ubyte(im),
    )
# %%


src/pattern_gen.py
#!/usr/bin/env python
# TODO Add documentation
# TODO Finish Type Hints
# TODO Release to pypi

__author__ = "Brad Day"
__email__ = "bradday4@gmail.com"
__license__ = "MIT"
import os
import sys
import pathlib
from functools import partial
from typing import Optional, Tuple, Union
import numpy as np
import fire
from scipy import signal
from PIL import Image

FILEFORMATS = frozenset(["tif", "tiff", "jpg", "jpeg", "bmp", "png"])
DEFAULTBAR = "sine"
BARTYPES = {
    "solid": signal.square,
    DEFAULTBAR: np.sin,
    "triangle": partial(signal.sawtooth, width=0.5),
}
ORIENTATION = frozenset(["vertical", "horizontal"])
DEVICEDEFAULTS = (1024, 1024, 8)
BITDEPTH = frozenset([8, 16, 32])
DTYPES = {8: np.uint8, 16: np.uint16, 32: np.uint32}
BITLIMS = {8: (0, 2 ** 8 - 1), 16: (0, 2 ** 16 - 1), 32: (0, 1)}
FORMATBITPAIRS = {
    "tif": frozenset([8, 16, 32]),
    "tiff": frozenset([8, 16, 32]),
    "jpg": frozenset([8]),
    "jpeg": frozenset([8]),
    "bmp": frozenset([8]),
    "png": frozenset([8, 16]),
}
EXTALIAS = {"tif": "tiff", "jpg": "jpeg"}
# lv: int,
# lh: int,
# wb: int,


def gen_pattern_image(sig, orientation, size) -> Image:
    if orientation == "horizontal":
        return Image.fromarray(np.tile(sig[:, np.newaxis], (1, size)))
    else:
        return Image.fromarray(np.tile(sig[np.newaxis, :], (size, 1)))


def rescale(x, lims=(0, 1)):
    new_min, new_max = lims
    old_min, old_max = np.min(x), np.max(x)
    return (new_max - new_min) * ((x - old_min) / (old_max - old_min)) + new_min


def cast(x: np.array, dtype: np.dtype):
    return x.astype(dtype)


def func_dispatcher(t: np.ndarray, bar_type: str):

    func = BARTYPES[bar_type]
    return func(t)


def verify_output_dir(output_directory: Union[str, None]) -> str:
    # check the directory exists
    if output_directory:
        if not os.path.exists(output_directory):
            raise ValueError(f"Output directory {output_directory} doesn't exist")
    else:
        if getattr(sys, "frozen", False):
            parent = os.path.dirname(sys.executable)
        elif __file__:
            parent = pathlib.Path(__file__).parent.absolute()

        output_directory = os.path.join(parent, "patterns")
        if not os.path.exists(output_directory):
            os.mkdir(output_directory)
        print(
            f"\nNo Output directory given using {output_directory} for patterned images\n"
        )
    return output_directory


class PatternGenerator:
    def __init__() -> None:
        # PatternGenerator class wraps runner method allowing import in other modules
        pass

    @staticmethod
    def runner(
        *args: Optional[Tuple[int, int, int]],
        bar_type: Optional[str] = "solid",
        orientation: str = "vertical",
        file_format: str = "bmp",
        phase_shifts: int = 3,
        output_directory: Optional[str] = None,
        bits: Optional[int] = 8,
        full_depth: Optional[bool] = True,
    ):
        # unpack the arguments:
        # if zero provided use defaults
        # if 3 are given unpack them else raise an error
        if args:
            if len(args) != 3:
                raise TypeError(f"Expected 3 arguments but {len(args)} were given")
            elif not all((isinstance(arg, int) for arg in args)):
                raise TypeError(f"Args must be of type int")
            lv, lh, wb = args
        else:
            lv, lh, wb = DEVICEDEFAULTS

        if bar_type not in BARTYPES:
            raise ValueError(
                f"Bar type must be one of {BARTYPES.keys()} but '{bar_type}' was given"
            )

        if orientation not in ORIENTATION:
            raise ValueError(
                f"Orientation must be either {ORIENTATION} but '{orientation}' was given"
            )

        # check file formats are recognized
        if file_format not in FILEFORMATS:
            raise ValueError(f"File format {file_format} not recognized")

        output_directory = verify_output_dir(output_directory)

        # check bit depth is correct
        if bits not in BITDEPTH:
            raise ValueError(f"Bit depth '{bits}' is not one of {BITDEPTH}")

        # check full_depth is of type bool
        if not isinstance(full_depth, bool):
            raise TypeError(
                f"flag full_depth must be True or False not type {type(full_depth)}"
            )
        # check full_depth is true if bartype is set to anything other than solid
        if bar_type != DEFAULTBAR and not full_depth:
            raise TypeError(f"bar_type {bar_type} only supported in full depth mode")

        # check chosen number of bits is compatabile with image format
        if not bits in FORMATBITPAIRS[file_format]:
            raise ValueError(
                f"File format {file_format} only takes bits {FORMATBITPAIRS[file_format]}"
            )

        # modify extensions
        if file_format in EXTALIAS:
            extension = file_format
            file_format = EXTALIAS[file_format]
        else:
            extension = file_format

        what_dim = lambda x: lh if x == "vertical" else lv
        tile_dim = lambda x: lv if x == "vertical" else lh
        freq = what_dim(orientation) / wb / 2

        phi_deg = np.linspace(0.001, 360, phase_shifts, endpoint=False)
        phi_rad = (np.radians(deg) for deg in phi_deg)  # phase in radian
        t = np.linspace(0, 1, what_dim(orientation), endpoint=False)  # spacing

        if full_depth:
            lims = BITLIMS[bits]
        else:
            lims = (0, 1)

        for i, phi in enumerate(phi_rad):

            sig = func_dispatcher(2 * np.pi * freq * t + phi, bar_type)
            sig = rescale(sig, lims=lims)
            sig = cast(sig, DTYPES[bits])

            image = gen_pattern_image(sig, orientation, tile_dim(orientation),)
            # x = np.sin(2 * np.pi * freq * t + phi)
            # x = rescale(x)
            # image = gen_pattern_image(x, orientation, what_dim(orientation))

            image.save(
                os.path.join(output_directory, f"phi_{i+1:02d}.{extension}"),
                file_format,
            )


if __name__ == "__main__":
    fire.Fire(PatternGenerator.runner)
#%%
# %%
